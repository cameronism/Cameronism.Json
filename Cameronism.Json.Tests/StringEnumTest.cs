using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace Cameronism.Json.Tests
{
	public unsafe class StringEnumTest
	{
		[Fact]
		public void DayOfWeekEnum()
		{
			var buffer = new byte[1024];
			int stringStart;
			int freeStart;

			fixed (byte* ptr = buffer)
			{
				var lookup = StringEnum.GenerateSequentialLookup(typeof(DayOfWeek), ptr, buffer.Length, out freeStart);

				Assert.True(lookup.HasValue, "lookup must have been generated");
				stringStart = (int)(lookup.Value.StringStart - ptr);
				
				//var chars = new char[64];
				//foreach (DayOfWeek dow in Enum.GetValues(typeof(DayOfWeek)))
				//{
				//	byte* str;
				//	int len;
				//	lookup.Value.FindSequential((ulong)dow, out str, out len);
				//	Assert.True(str != null, "string must not be null");
				//	var dowString = dow.ToString();
				//	Assert.Equal(dowString.Length, len);

				//	string actualString;
				//	fixed (char* ch = chars)
				//	{
				//		var strLen = Encoding.UTF8.GetChars(str, len, ch, chars.Length);
				//		actualString = new string(ch, 0, strLen);
				//	}
				//	Assert.Equal(dowString, actualString);
				//}
			}

			var sb = new StringBuilder();
			sb.AppendLine("## Lookups");
			Util.Hex.Dump(sb, buffer.Take(stringStart));
			sb.AppendLine();
			sb.AppendLine("## Strings");
			Util.Hex.Dump(sb, buffer.Skip(stringStart).Take(freeStart - stringStart));

			ApprovalTests.Approvals.Verify(sb.ToString());
		}
	}
}
