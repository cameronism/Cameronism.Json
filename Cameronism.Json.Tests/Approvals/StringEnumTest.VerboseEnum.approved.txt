# Verbose System.DayOfWeek
## Lookups


0000   00 00 00 00 00 00 00 00   00 00 53 75 6e 64 61 79   ........ ..Sunday
0010   01 00 00 00 00 00 00 00   00 00 4d 6f 6e 64 61 79   ........ ..Monday
0020   02 00 00 00 00 00 00 00   00 54 75 65 73 64 61 79   ........ .Tuesday
0030   03 00 00 00 00 00 00 00   80 00 09 00 00 00 00 00   ........ ........
0040   04 00 00 00 00 00 00 00   54 68 75 72 73 64 61 79   ........ Thursday
0050   05 00 00 00 00 00 00 00   00 00 46 72 69 64 61 79   ........ ..Friday
0060   06 00 00 00 00 00 00 00   53 61 74 75 72 64 61 79   ........ Saturday

## Strings


0000   57 65 64 6e 65 73 64 61   79                        Wednesda y

# Verbose Cameronism.Json.Tests.StringEnumTest+Sparse
## Lookups


0000   01 00 00 00 00 00 00 00   00 00 00 00 00 00 76 31   ........ ......v1
0010   02 00 00 00 00 00 00 00   00 00 00 00 00 00 76 32   ........ ......v2
0020   04 00 00 00 00 00 00 00   00 00 00 00 00 00 76 33   ........ ......v3
0030   08 00 00 00 00 00 00 00   00 00 00 00 00 00 76 34   ........ ......v4
0040   10 00 00 00 00 00 00 00   00 00 00 00 00 00 76 35   ........ ......v5
0050   20 00 00 00 00 00 00 00   00 00 00 00 00 00 76 36    ....... ......v6
0060   40 00 00 00 00 00 00 00   00 00 00 00 00 00 76 37   @....... ......v7
0070   80 00 00 00 00 00 00 00   00 00 00 00 00 00 76 38   ........ ......v8
0080   00 01 00 00 00 00 00 00   00 00 00 00 00 00 76 39   ........ ......v9
0090   00 02 00 00 00 00 00 00   00 00 00 00 00 76 31 30   ........ .....v10
00a0   00 04 00 00 00 00 00 00   00 00 00 00 00 76 31 31   ........ .....v11
00b0   00 08 00 00 00 00 00 00   00 00 00 00 00 76 31 32   ........ .....v12
00c0   00 10 00 00 00 00 00 00   00 00 00 00 00 76 31 33   ........ .....v13
00d0   00 20 00 00 00 00 00 00   00 00 00 00 00 76 31 34   . ...... .....v14
00e0   00 40 00 00 00 00 00 00   00 00 00 00 00 76 31 35   .@...... .....v15
00f0   00 80 00 00 00 00 00 00   00 00 00 00 00 76 31 36   ........ .....v16
0100   00 00 01 00 00 00 00 00   00 00 00 00 00 76 31 37   ........ .....v17
0110   00 00 02 00 00 00 00 00   00 00 00 00 00 76 31 38   ........ .....v18
0120   00 00 04 00 00 00 00 00   00 00 00 00 00 76 31 39   ........ .....v19
0130   00 00 08 00 00 00 00 00   00 00 00 00 00 76 32 30   ........ .....v20
0140   00 00 10 00 00 00 00 00   00 00 00 00 00 76 32 31   ........ .....v21
0150   00 00 20 00 00 00 00 00   00 00 00 00 00 76 32 32   .. ..... .....v22
0160   00 00 40 00 00 00 00 00   00 00 00 00 00 76 32 33   ..@..... .....v23
0170   00 00 80 00 00 00 00 00   00 00 00 00 00 76 32 34   ........ .....v24
0180   00 00 00 01 00 00 00 00   00 00 00 00 00 76 32 35   ........ .....v25
0190   00 00 00 02 00 00 00 00   00 00 00 00 00 76 32 36   ........ .....v26
01a0   00 00 00 04 00 00 00 00   00 00 00 00 00 76 32 37   ........ .....v27
01b0   00 00 00 08 00 00 00 00   00 00 00 00 00 76 32 38   ........ .....v28
01c0   00 00 00 10 00 00 00 00   00 00 00 00 00 76 32 39   ........ .....v29
01d0   00 00 00 20 00 00 00 00   00 00 00 00 00 76 33 30   ... .... .....v30
01e0   00 00 00 40 00 00 00 00   00 00 00 00 00 76 33 31   ...@.... .....v31
01f0   00 00 00 80 00 00 00 00   00 00 00 00 00 76 33 32   ........ .....v32

## Strings

zero bytes dumped

# Verbose Cameronism.Json.Tests.StringEnumTest+Skippy
## Lookups


0000   01 00 00 00 00 00 00 00   00 00 00 00 00 46 6f 6f   ........ .....Foo
0010   03 00 00 00 00 00 00 00   00 00 00 00 00 42 61 72   ........ .....Bar
0020   05 00 00 00 00 00 00 00   00 00 00 00 00 42 6f 70   ........ .....Bop
0030   07 00 00 00 00 00 00 00   00 00 00 00 00 42 61 7a   ........ .....Baz

## Strings

zero bytes dumped

# Verbose Cameronism.Json.Tests.StringEnumTest+Negatives
## Lookups


0000   01 00 00 00 00 00 00 00   00 00 00 00 00 4f 6e 65   ........ .....One
0010   ff ff ff 7f 00 00 00 00   00 00 00 00 00 4d 61 78   ........ .....Max
0020   00 00 00 80 ff ff ff ff   00 00 00 00 00 4d 69 6e   ........ .....Min
0030   fe 7f ff ff ff ff ff ff   80 00 08 00 00 00 00 00   ........ ........
0040   ff 7f ff ff ff ff ff ff   00 ef bd b2 e3 82 93 61   ........ .......a
0050   ff ff ff ff ff ff ff ff   80 00 0b 00 08 00 00 00   ........ ........

## Strings


0000   ef bd b2 e3 82 93 61 61   4e 65 67 61 74 69 76 65   ......aa Negative
0010   4f 6e 65                                            One